File: Data Structures & Algorithms/find-target-in-rotated-sorted-array/submission-3.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        n = len(nums)
        l,r= 0,n-1
        if nums[l]<nums[r]:
            while l<=r:
                mid = (l+r)//2
                if target == nums[mid]:
                    return mid
                elif target<nums[mid]:
                    r = mid-1
                else:
                    l = mid+1
            return -1
        while l<r:
            mid = (l+r)//2
            if nums[mid]>nums[r]:
                l = mid+1
            else:
                r = mid
        print(f'{l}:{nums[l]}, target: {target}')
        if nums[l]==target:
            return l
        if nums[l]<=target<=nums[n-1]:
            print(f'{l}:{nums[l]}, target: {target}')
            print(" nums[l] < target < nums[n-1] ")
            low = l+1
            high = n-1
            while low<=high:
                mid = (low+high)//2
                if target == nums[mid]:
                    return mid
                elif target<nums[mid]:
                    high = mid-1
                else:
                    low = mid+1
        else:
            low = 0
            high = l-1
            while low<=high:
                mid = (low+high)//2
                if target == nums[mid]:
                    return mid
                elif target<nums[mid]:
                    high = mid-1
                else:
                    low = mid+1
        return -1

File: Data Structures & Algorithms/longest-increasing-subsequence/submission-3.java
class Solution {
    public int lengthOfLIS(int[] nums) {
        int n = nums.length;
        ArrayList<Integer> tails = new ArrayList<>();

        for (int num : nums) {
            int left = 0;
            int tail_length = tails.size();
            int right = tail_length - 1;

            while (left <= right) {
                int mid = left + (right - left) / 2;

                if (tails.get(mid) < num) {
                    left = mid + 1;
                } else {
                    right = mid - 1;
                }
            }
            if (left == tail_length) {
                tails.add(num);
            } else {
                tails.set(left, num);
            }
        }
        return tails.size();
    }
}